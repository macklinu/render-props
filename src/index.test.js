import React from 'react'
import renderer from 'react-test-renderer'
import renderProps from '.'

const Counter = props => renderProps(props, { count: 1 })
const Component = ({ count }) => <div>{count}</div>
const render = element => renderer.create(element).toJSON()

test('accepts component prop', () => {
  expect(
    render(<Counter component={({ count }) => <div>{count}</div>} />)
  ).toMatchSnapshot()
  expect(render(<Counter component={Component} />)).toMatchSnapshot()
})

test('accepts render prop', () => {
  expect(
    render(<Counter render={({ count }) => <div>{count}</div>} />)
  ).toMatchSnapshot()
  expect(render(<Counter render={Component} />)).toMatchSnapshot()
})

test('accepts children function prop', () => {
  expect(
    render(<Counter>{({ count }) => <div>{count}</div>}</Counter>)
  ).toMatchSnapshot()
  expect(render(<Counter>{Component}</Counter>)).toMatchSnapshot()
})

test('renders component prop above all other props passed to it', () => {
  const createMockComponent = () => jest.fn(props => <Component {...props} />)
  const MockComponent = createMockComponent()
  const RenderComponent = createMockComponent()
  const ChildrenComponent = createMockComponent()

  render(
    <Counter component={MockComponent} render={RenderComponent}>
      {ChildrenComponent}
    </Counter>
  )

  expect(MockComponent).toHaveBeenCalled()
  expect(RenderComponent).not.toHaveBeenCalled()
  expect(ChildrenComponent).not.toHaveBeenCalled()
})

test('renders null if render functions are not passed to component', () => {
  expect(render(<Counter />)).toBeNull()
})

test('throws if no component props are passed', () => {
  const InvalidComponent = _props => renderProps(null, { any: 'any' })

  expect(() => render(<InvalidComponent />)).toThrowErrorMatchingSnapshot()
})

test('throws if no new props are passed', () => {
  const InvalidComponent = props => renderProps(props)

  expect(() => render(<InvalidComponent />)).toThrowErrorMatchingSnapshot()
})

test('throws if component props is truthy but not an object', () => {
  const InvalidComponent = _props => renderProps('nope', { any: 'any' })

  expect(() => render(<InvalidComponent />)).toThrowErrorMatchingSnapshot()
})

test('throws if new props are truthy but not an object', () => {
  const InvalidComponent = props => renderProps(props, 'nope')

  expect(() => render(<InvalidComponent />)).toThrowErrorMatchingSnapshot()
})
