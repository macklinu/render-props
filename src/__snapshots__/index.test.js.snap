// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accepts children function prop 1`] = `
<div>
  1
</div>
`;

exports[`accepts children function prop 2`] = `
<div>
  1
</div>
`;

exports[`accepts component prop 1`] = `
<div>
  1
</div>
`;

exports[`accepts component prop 2`] = `
<div>
  1
</div>
`;

exports[`accepts render prop 1`] = `
<div>
  1
</div>
`;

exports[`accepts render prop 2`] = `
<div>
  1
</div>
`;

exports[`forwarded props can be of any type 1`] = `
<pre>
  {"hello":"world"}
</pre>
`;

exports[`throws if component props is truthy but not an object 1`] = `"Must supply truthy component props"`;

exports[`throws if no component props are passed 1`] = `"Must supply truthy component props"`;
